internals-js/src/directiveCompositionRules.ts

```diff
@@ -0,0 +1,255 @@
+import { DirectiveLocation } from 'graphql';
+import { Directive, DirectiveDefinition, FieldDefinition, NamedSchemaElement, ObjectType, Schema } from './definitions'
+import { assert, isDefined, isNotNull } from './utils';
+
+export enum DirectiveCompositionStrategy {
+  COLLAPSE = 'collapse',
+  COLLAPSE_FROM_ALL = 'collapseFromAll',
+  REMOVE_DUPLICATES = 'removeDuplicates',
+}
+
+export enum DirectivePropagationStrategy {
+  INHERIT_FROM_OBJECT = 'inheritFromObject',
+  CONSISTENT_LOCATION = 'consistentLocation',
+}
+
+export enum FieldPropagationStrategy {
+  MAX = 'max',
+  MIN = 'min',
+  SUM = 'sum',
+  AVERAGE = 'average',
+  AND = 'and',
+  OR = 'or',
+  INTERSECTION = 'intersection',
+  UNION = 'union',
+  EXACT = 'exact',
+}
+
+const SUPPORTED_LOCATIONS = [
```

Comment:
should this include some other locations? e.g. interface, enums etc?

Reply:
My thought was that we'd expand support on an as-needed basis. For the stuff that's upcoming, I think OBJECT and FIELD_DEFINITION is sufficient.

Reply:
The cost directive will be on  ARGUMENT_DEFINITION, ENUM, FIELD_DEFINITION, INPUT_FIELD_DEFINITION, OBJECT and SCALAR. Does that make a difference here?

Reply:
Yeah, saw that once you sent me the definition. I'll make the changes to support those types as well.

Reply:
Mainly so that if someone tries to add a directive with an unsupported location, they'll know there's a problem right away rather than assuming it works.

Reply:
But what "problem" are we talking about here? Why would any location be a problem for any of this?